packages/react-components/src/components/TextFieldWithMention/TextFieldWithMention.tsx

```diff
@@ -0,0 +1,629 @@
+          // the selection end is changed
+          const newIndex = getUpdatedSelectionStartIndexOrEndIndex(
+            inputTextValue,
+            tagsValue,
+            currentSelectionEnd,
+            selectionEndValue
+          );
+          updatedEndIndex = newIndex !== undefined ? newIndex : updatedEndIndex;
+        }
+      }
+      // e.currentTarget.selectionDirection should be set to handle shift + arrow keys
+      if (event.currentTarget.selectionDirection === null) {
+        event.currentTarget.setSelectionRange(
+          updatedStartIndex === undefined ? null : updatedStartIndex,
+          updatedEndIndex === undefined ? null : updatedEndIndex
+        );
+      } else {
+        event.currentTarget.setSelectionRange(
+          updatedStartIndex === undefined ? null : updatedStartIndex,
+          updatedEndIndex === undefined ? null : updatedEndIndex,
+          event.currentTarget.selectionDirection
+        );
+      }
+      setSelectionStartValue(updatedStartIndex);
+      setSelectionEndValue(updatedEndIndex);
+    },
+    [getUpdatedSelectionStartIndexOrEndIndex]
+  );
+
+  const handleOnSelect = useCallback(
+    (
+      event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
+      inputTextValue: string,
+      tags: TagData[],
+      shouldHandleOnMouseDownDuringSelect: boolean,
+      selectionStartValue?: number,
+      selectionEndValue?: number
+    ): void => {
+      const currentSelectionStart =
+        event.currentTarget.selectionStart === null ? undefined : event.currentTarget.selectionStart;
+      const currentSelectionEnd =
+        event.currentTarget.selectionEnd === null ? undefined : event.currentTarget.selectionEnd;
+      if (shouldHandleOnMouseDownDuringSelect && currentSelectionStart !== undefined) {
+        // on select was triggered by mouse down
+        const mentionTag = findMentionTagForSelection(tags, currentSelectionStart);
+        if (mentionTag !== undefined && mentionTag.plainTextBeginIndex !== undefined) {
+          // handle mention click
+          if (event.currentTarget.selectionDirection === null) {
+            event.currentTarget.setSelectionRange(
+              mentionTag.plainTextBeginIndex,
+              mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex
+            );
+          } else {
+            event.currentTarget.setSelectionRange(
+              mentionTag.plainTextBeginIndex,
+              mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex,
+              event.currentTarget.selectionDirection
+            );
+          }
+          setSelectionStartValue(mentionTag.plainTextBeginIndex);
+          setSelectionEndValue(mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex);
+        } else {
+          setSelectionStartValue(currentSelectionStart);
+          setSelectionEndValue(currentSelectionEnd);
+        }
+      } else {
```

Comment:
Feeling it would be more readable if we could do rename + extract some code lines into functions like
if(...) {
   handleSelectionByMouseDown();
} else {
  handleSelectionByKeyboard();
}

Comments would work very well for sure here, but it would be always better if the code structure can make it self-commented